lib/auth/userloginstate/generator.go

```diff
@@ -431,10 +431,36 @@ func (g *Generator) Refresh(ctx context.Context, user types.User, ulsService ser
 	user.SetRoles(uls.GetRoles())
 	user.SetTraits(uls.GetTraits())
 
-	uls, err = ulsService.UpsertUserLoginState(ctx, uls)
+	// TODO(smallinsky) In case of unhealthy cached in best case we will still do 1 api call in worst case 2 api calls.
+	// So this optimization make sense only if cache is healthy.
+	// Right now cache interface is not exposing the healthy state.
+	// We should rework this and do 1 Upsert call is cache is down.
+	uls, err = maybeUpdateUserLoginState(ctx, ulsService, uls)
```

Comment:
This optimization only makes sense if we are reading from the actual storage, the cache can be healthy and stil be lagging behind (and if we're deliberately reading from actual storage this comment doesn't apply, I think).